src/pgwire/src/protocol.rs

```diff
@@ -233,6 +237,7 @@ where
             metrics,
             conn,
             coord_client: &mut coord_client,
+            frontegg_check,
         };
         machine.run().await
```

Comment:
Even better, I think that means you can do this without a separate Tokio task that has complicated cancellation semantics. I think the implementation of `is_expired` can just do:

1. Sleep for `expiration - check_before_secs`.
2. Select on reaching the expiration time or refreshing the token.
3. If reached the expiration time, resolve with error.
4. If refreshing the token failed, resolve with error.
5. Otherwise, go to step 1.

Reply:
If the connection closes normally, will rust dropping this `is_expired` future result in also dropping any http requests and other stuff it's doing, like the above loop?

Reply:
Yep. It's usually one of the more painful and unexpected parts of `async` in Rust, but it's exactly the semantics we want here. Dropping the future _hard_ cancels all outstanding work.